content/blog/2024/proxy-store.md

```diff
@@ -0,0 +1,174 @@
+
+Recently, I was building the Playground feature on the [Sass
+website](https://sass-lang.com/playground/). It's an interactive feature built
+into a static Eleventy site, so we didn't have a UI Framework already
+integrated.
+
+{{ embed.img(
+  src='blog/2024/sass-playground.png',
+  alt='A screenshot of the Sass playground web app.'
+) }}
+
+Let's take a look at the interactive parts of the tool. Users have three inputs:
+
+- Input format radio (SCSS or Sass)
+- Output format radio (Expanded or Compressed)
+- Input text value (the input styles)
+
+From this, there are several side effects. Any time one of these changes, we
+need to recompile the Sass. We also need to update the hash in the URL, which
+allows users to share the state of the page.
+
+If compilation succeeds, we can simply update the output value. If it fails, we
+need to display the error.
+
+While this isn't an extremely complex set of interactions, it's enough that we
+want to be mindful of how we set this up. Essentially, we have three inputs and
+three outputs.
+
+## Defining what we need
+
+So, what is it that we actually need here?
+
+1. We need a central place to hold our data, or our "state."
+2. When our data changes, we need a way to trigger the appropriate side effects.
+
+## Finding a solution that fits
+
+At this point, there are a few options. A UI Framework like
+[Svelte](https://svelte.dev/) would provide an easy, centralized interface for
+reactivity, but comes with the long term update and maintenance costs of an
+extra dependency.
+
+We could also add our logic directly in click handlers of our inputs. This is
+perhaps the most straightforward approach, but also is decentralized, and would
+come with a lot of code duplication.
+
+A third solution is using a
+[Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy).
+We can store our data in a proxied object, and then intercept -- or "trap" --
+the setters to trigger the reactivity.
+
+## A simple Proxy state
+
+```js
+const initialState = { input: '' };
+
+const state = new Proxy(initialState, {
+  set(state, prop, value) {
+    const set = Reflect.set(...arguments);
+    console.log(`Setting ${prop} to ${value}`);
+    return set;
+  },
+});
+
+state.input = 'Hello!';
+// "Setting input to Hello!"
+```
+
+Every time we set the value of `state.input`, we log the prop and value. Note
+that we use the proxied object `state`, and not the `initialState` object.
+Because we are trapping `set`, we need to set the value with `Reflect.set()`. We
+also need to return a Boolean specifying whether the `set` was successful.
+
+## A more complex Proxy state
+
+```js
+const initialState = {
+  inputFormat: 'scss',
+  outputFormat: 'expanded',
+  inputValue: '',
+  compilerHasError: false,
+};
+
+const playgroundState = new Proxy(initialState, {
+  set(state, prop, ...rest) {
+    // Set state first so called functions have access
+    const set = Reflect.set(state, prop, ...rest);
```

Comment:
Is there a reason to specify the arguments to `Reflect.set` here? I think I'd expect:

```suggestion
  set(state, prop, value) {
    // Set state first so called functions have access
    const set = Reflect.set(...arguments);
```

Reply:
We used that pattern for types in the playground, but since this pulled out the types, there's no reason here.